src/nbl/video/CVulkanLogicalDevice.h

```diff
@@ -1304,7 +1308,7 @@ class CVulkanLogicalDevice final : public ILogicalDevice
         VkPipelineLayoutCreateInfo vk_createInfo = { VK_STRUCTURE_TYPE_PIPELINE_LAYOUT_CREATE_INFO };
         vk_createInfo.pNext = nullptr; // pNext must be NULL
         vk_createInfo.flags = static_cast<VkPipelineLayoutCreateFlags>(0); // flags must be 0
-        vk_createInfo.setLayoutCount = dsLayoutCount;
+        vk_createInfo.setLayoutCount = asset::ICPUPipelineLayout::DESCRIPTOR_SET_COUNT;
```

Comment:
huh?

Reply:
Judging by where it lives and what it is hardcoded to, `DESCRIPTOR_SET_COUNT` is a misleading name it should be
`DESCRIPTOR_SET_LAYOUT_COUNT`.

Dummy ds layouts are required to use `set = N` in the shader (without using all previous N-1 sets).

Reply:
yeah but you're always creating a pipeline layout with 4 descriptor set layouts, whereas you should be creating it with maximum non null layout count

Reply:
Let's take an example, for the decoration:
`layout (set = 3, binding = 0)`
you want:
- 1 DS Layout
- 4 DS (first 3 dummy and last one from the above Layout)
- Pipeline layout with the 1 DS layout from above
then you want to do: `vkCmdBindDescriptorSets(.., 3, 1, ..)`
Is that right?

Reply:
The Vulkan pipeline layout you create will have to have first 3 layouts refer to the same empty layout (created at logical device init), then the 4th one would be the non-nullptr layout passed into the function from outside.

You can track the bound pipelines and descriptor sets in the commandbuffer during record (after all there's only 4 things to track, and there's a common high level function in `IGPUPipelineLayout` or `IPipelineLayout` for checking pipeline compatibility rules which tells you how many descriptor sets will get unbound after changing a pipeline layout).

Just before any draw or dispatch, check whether all descriptors bound below **the maximum set number used by the pipeline layout of the pipeline currently bound** are non-null, if any are null then bind the empty descriptor in its place ( I guess that would require having the empty descriptor as an `IGPUDescriptorSet` rather than just a simple `vkDescriptorSet` handle).

Reply:
Resolved.